hack/build-antrea-windows-all.sh

```diff
@@ -27,7 +27,12 @@ _usage="Usage: $0 [--pull] [--push-base-images]
 Build the antrea/antrea-windows image, as well as all the base images in the build chain. This is
 typically used in CI to build the image with the latest version of all dependencies, taking into
 account changes to all Dockerfiles.
-        --pull                  Always attempt to pull a newer version of the base images.
+        --pull                  Always attempt to pull a newer version of the base images, ignoring pull-mask.
+        --pull-mask=mask        Binary mask to determine which base images to pull.
+                                  0x1 - antrea/windows-utility-base
```

Comment:
May I know what are `antrea/windows-utility-base` and `antrea/windows-golang` ? I didn't remember we used to publish these images. So where did we pull them?

Reply:
They are used in build/images/base-windows/Dockerfile, we pull them in hack/build-antrea-windows-all.sh.

Reply:
They (intermediate) must have been pushed for caching purposes. However, that never worked as expected I think. This is why we are using `docker buildx` on Linux now, so we can use the registry cache backend (https://docs.docker.com/build/cache/backends/), with a dedicated "cache image". I would recommend trying to do the same thing for the Windows build chain, because the changes in this PR (I haven't done an in-depth review) seem a bit too manual and error-prone. Our build on Linux is quite fast now because the caching works very well IMO. Even with some code changes, it only takes 3-4 minutes (in Github CI).

See for example: https://github.com/antrea-io/antrea/blob/20ca6503b10ea5440444474a427dea3a3f2e250c/build/images/base/build.sh#L138-L156

Reply:
> They (intermediate) must have been pushed for caching purposes. However, that never worked as expected I think. This is why we are using `docker buildx` on Linux now, so we can use the registry cache backend (https://docs.docker.com/build/cache/backends/), with a dedicated "cache image". I would recommend trying to do the same thing for the Windows build chain, because the changes in this PR (I haven't done an in-depth review) seem a bit too manual and error-prone. Our build on Linux is quite fast now because the caching works very well IMO. Even with some code changes, it only takes 3-4 minutes (in Github CI).
> 
> See for example:
> 
> https://github.com/antrea-io/antrea/blob/20ca6503b10ea5440444474a427dea3a3f2e250c/build/images/base/build.sh#L138-L156

@antoninbas Thanks for the clarification, I still have one question. I see that Image `windows-utility-base` can speed up the installation of related libraries, but I am unclear on the benefits of `windows-golang`. From the Dockerfile https://github.com/antrea-io/antrea/blob/52515028cb273bc54abbc5c2a53cc345e2a51737/build/images/base-windows/Dockerfile#L46 , I don’t understand what the difference is between using `windows-golang` and directly pulling Image `golang:${GO_VERSION}-nanoserver`.

After determining all the necessary intermediate images, I will introduce Windows Docker Buildx in this PR.

Reply:
I think this was introduced by you in https://github.com/antrea-io/antrea/pull/3518?

I am not sure why we have it. `COPY --from=golang:${GO_VERSION}-nanoserver` should also work according to https://docs.docker.com/build/building/multi-stage/#use-an-external-image-as-a-stage.

Reply:
> I think this was introduced by you in #3518?
> 
> I am not sure why we have it. `COPY --from=golang:${GO_VERSION}-nanoserver` should also work according to https://docs.docker.com/build/building/multi-stage/#use-an-external-image-as-a-stage.

I mean can we remove `windows-golang` from intermediate images we built in build-antrea-windows-all.sh? https://github.com/antrea-io/antrea/blob/d4922f06f75a48572d8e8bcb2daf596524ab3f7e/hack/build-antrea-windows-all.sh#L95-L100

If so, we can maintain only the three intermediate images (windows-utility-base, base-windows, windows-ovs) for the Antrea Windows image build.

Reply:
This code was just doing the same thing we were doing on Linux. All intermediate images were built separately and pushed to the registry separately, so they could be pulled for subsequent builds (including on different machines) for caching purposes. There is some context in this old issue: https://github.com/moby/moby/issues/34715. This intermediate image (`windows-golang`) is a bit special, since we won't have any build step besides the `FROM` statement in the Dockerfile, but I guess we just "blindly" followed the same steps for all intermediate images in all Dockerfiles. Note that this approach stopped working at some point (assuming it ever worked). We are no longer doing this for Linux now that we use BuildKit and we export the build cache to the registry as a separate image. This works quite well.

In this specific case, it feels like there isn't even a need for the intermediate image (`windows-golang`), not just in the registry but also locally, as we could just do `COPY --from=golang:${GO_VERSION}-nanoserver ...`.

Because the approach described above (pushing intermediate targets to the registry for caching) doesn't work anyway, there should not be any value in pushing `windows-utility-base` either. This could be confirmed by looking at a few CI job logs.